packages/workflow/src/CredentialSchema/CredentialSchema.ts

```diff
@@ -0,0 +1,274 @@
+import z, { type ZodOptional, type ZodType } from 'zod';
+
+import type { INodeProperties } from '../Interfaces';
+
+function isObject(value: unknown): value is object {
+	return typeof value === 'object' && value !== null && !Array.isArray(value);
+}
+
+function removeUndefinedProperties<T extends object>(obj: T): T {
+	for (const [key, value] of Object.entries(obj)) {
+		if (value === undefined) {
+			delete (obj as Record<string, unknown>)[key];
+		} else if (isObject(value)) {
+			removeUndefinedProperties(value);
+			if (Object.keys(value).length === 0) {
+				delete (obj as Record<string, unknown>)[key];
+			}
+		}
+	}
+	return obj;
+}
+
+class CredentialSchemaRootObject<
+	M extends BaseMetadata,
+	S extends ZodType | null,
+	T extends { [k: string]: CredentialSchemaProperty<M, S> } = {},
+> {
+	constructor(public shape: T) {}
+
+	validate<Data>(data: Data) {
+		return this.toZodSchema().safeParse(data);
+	}
+
+	toZodSchema() {
+		return z.object(
+			Object.fromEntries(
+				Object.entries(this.shape)
+					.filter(([_, property]) => property.toZodSchema())
```

Comment:
Found it a bit unintuitive that we're filtering like this, which may return a falsy value.